File: DailyChallenges/CPP/LeetCode_926.cpp
class Solution {
public:
    int minFlipsMonoIncr(string s) {
        int n = s.size();
        vector<int> flip_0(n+1); // 1 to n
        for (int i=1; i<=n; ++i)
            flip_0[i] = flip_0[i-1] + ((s[i-1] == '0') ? 0 : 1);

        int ans = flip_0[n]; // all flip to 0
        int flip_1 = 0;
        for (int i = n-1; i>=0; --i){
            flip_1 += (s[i] == '1') ? 0 : 1;
            ans = min(ans, flip_1 + flip_0[i]);
        }

        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_926.scala
object Solution {
    def minFlipsMonoIncr(s: String): Int = {
        val f0 = s.scanLeft(0)((cnt, ch) => if (ch == '0') cnt else cnt+1).tail
        val f1 = s.scanRight(0)((ch, cnt) => if (ch == '1') cnt else cnt+1).init
        (f0 zip f1).map(p => p._1 + p._2 -1).min
    }
}
